File: python/problem_6.py
# xorNode stores a value and the XOR of its neighbours' indices.
# A list of nodes backed by a Python list simulates pointer arithmetic.

class XorLinkedList:
    # XorLinkedList is an XOR doubly-linked list backed by a Python list.
    # Using list indices as stand-ins for memory addresses.
    def __init__(self):
        self._nodes = []  # list of (val, both) tuples; index = "address"
        self._head = -1
        self._tail = -1

    def add(self, val: int) -> None:
        # append element to the end of the list
        idx = len(self._nodes)
        prev = self._tail
        self._nodes.append([val, prev ^ -1])  # both = prev XOR null(-1)
        if self._tail != -1:
            # update old tail: replace null(-1) with new index
            old_val, old_both = self._nodes[self._tail]
            self._nodes[self._tail][1] = old_both ^ (-1 ^ idx)
        self._tail = idx
        if self._head == -1:
            self._head = idx

    def get(self, index: int) -> int:
        # traverse to the index-th node (0-based)
        prev, curr = -1, self._head
        for _ in range(index):
            nxt = prev ^ self._nodes[curr][1]  # XOR out prev to get next
            prev, curr = curr, nxt
        return self._nodes[curr][0]


def main():
    print("Problem 6:")
    lst = XorLinkedList()
    for v in [1, 2, 3, 4]:
        lst.add(v)
    assert lst.get(0) == 1, "assertion failed"
    assert lst.get(1) == 2, "assertion failed"
    assert lst.get(2) == 3, "assertion failed"
    assert lst.get(3) == 4, "assertion failed"
    print("All Tests Passed")

File: lua/problem_6.lua
-- XorLinkedList is an XOR doubly-linked list backed by a Lua table.
-- Table indices (1-based) serve as stand-in memory addresses; -1 is null.

local XorLinkedList = {}
XorLinkedList.__index = XorLinkedList

local function newXorLinkedList()
    return setmetatable({ nodes = {}, head = -1, tail = -1 }, XorLinkedList)
end

-- add appends an element to the end of the list.
function XorLinkedList:add(val)
    local idx = #self.nodes + 1   -- 1-based index
    local prev = self.tail
    self.nodes[idx] = { val = val, both = prev ~ -1 }  -- both = prev XOR null(-1)
    if self.tail ~= -1 then
        -- update old tail: replace null(-1) with new index
        self.nodes[self.tail].both = self.nodes[self.tail].both ~ (-1 ~ idx)
    end
    self.tail = idx
    if self.head == -1 then self.head = idx end
end

-- get returns the value at the given 0-based index.
function XorLinkedList:get(index)
    local prev, curr = -1, self.head
    for _ = 1, index do
        local nxt = prev ~ self.nodes[curr].both  -- XOR out prev to get next
        prev, curr = curr, nxt
    end
    return self.nodes[curr].val
end

local function problem6()
    print("Problem 6:")
    local lst = newXorLinkedList()
    for _, v in ipairs({1, 2, 3, 4}) do lst:add(v) end
    assert(lst:get(0) == 1, "assertion failed")
    assert(lst:get(1) == 2, "assertion failed")
    assert(lst:get(2) == 3, "assertion failed")
    assert(lst:get(3) == 4, "assertion failed")
    print("All Tests Passed")
end

return { problem6 = problem6 }
